### ISSUE TITLE
Add a home-made termination checker for fixpoints

### ISSUE THREAD

**Issue by strub** [EasyCrypt/easycrypt#144]
Currently, fixpoints are checked for termination by Why3 when a goal is sent to SMT solvers. This is clearly a silly behvior.

**Comment by oskgo** [comment]
Isn't this is a duplicate of #62?

**Comment by chdoc** [comment]
No, IIUC then #62 is about what is accepted as terminating. The issue here is when termination is checked. Currently, EC seems to accept obvious nonsense such as:
```
op foo (n : nat) = 
  with n = O => 0
  with n = S n' => S (foo (S n')).
```
Fortunately, there doesn't seem to be a way to unfold a recursive definition by one step. If there is, then I'm pretty sure this is a soundness bug:
```
lemma oops n : foo (S n) = S (foo (S n)). 
(* rewrite {1}/foo. -> the operator cannot be unfolded *) 
(* rewrite /=       -> anomaly: Stack overflow *
(* done             -> anomaly: Stack overflow *) 
(* smt()            -> anomaly: Stack overflow *)
```

**Comment by strub** [comment]
As said, EC does not check for termination at all - only Why3 is doing the check. The question is which kind of termination check we want. We can implement a very simple syntactical check.

But anyway, not that we *cannot* avoid Why3 doing the check on its side. This means that either we add a flag to the Why3 API to disable the check, or we do not use the Why3 API for fixpoints and encode them using an opaque definition + axioms.

**Comment by strub** [comment]
> @strub , I think before implementing a home-made termination checker, we should check termination at the time of definition rather than at the time of use.

The definition is not checked when used on purpose. It is checked by Why3 when doing a call to `smt` because the tactics is doing a translation to Why3.

**Comment by strub** [comment]
And, to be honest, I don't think that the Why3 termination checker does anything else than checking that you the inductive argument decreases w.r.t. the (lexicographic-extension) of the subterm ordering.

**Comment by chdoc** [comment]
Taking the original example from #62 , it appears to also have - similar to Coq - the notion of "structurally less than or equal" and establish this for (previously defined) recursive definitions applied to the decreasing argument. Talk to a Coq dev about the guard condition, and I promise you material for nightmares. 

> The definition is not checked when used on purpose. It is checked by Why3 when doing a call to `smt` because the tactics is doing a translation to Why3.

I don't quite understand. Am I to read this as "recursive definitions are only known to the SMT and will not be reasoned about without SMT". If we add a tactic that unfolds a recursive definition once, then I don't need to call SMT to prove false.

What's keeping us from triggering the termination check from Why3 at the time a recursive definition is stated?

**Comment by chdoc** [comment]
Thinking further on this, how about asking SMT to immediately prove the "unfolding rules" (e.g. the `oops` above) for any recursive definition? These are anyway useful, and this should fail for definitions that Why3 cannot prove terminating. Naming these may be an issue, but `<op-name>_eq<n>` doesn't look too bad to me.
